app/uk/gov/hmrc/agentsubscriptionfrontend/controllers/SubscriptionController.scala

```diff
@@ -113,28 +106,40 @@ class SubscriptionController @Inject()
     implicit authContext =>
       implicit request =>
 
-        import SubscriptionDetails._
         import Address._
+        import SubscriptionDetails._
+
+        def subscribe(details: InitialDetails, address: Address): Future[Either[Int, (Arn, String)]] = {
+          val subscriptionDetails = mapper(details, address)
+          subscriptionService.subscribeAgencyToMtd(subscriptionDetails) map {
+            case Right(arn) => Right((arn, subscriptionDetails.name))
+            case Left(x) => Left(x)
+          }
+        }
 
         addressLookUpConnector.getAddressDetails(id).flatMap { address =>
           validate(address, blacklistedPostCodes) match {
             case Invalid(errors) =>
-
-              Future {
+              Future.successful(
                 Ok(uk.gov.hmrc.agentsubscriptionfrontend.views.html.des_will_not_accept_address(id, renderErrors(errors)))
-              }
+              )
             case Valid(()) =>
-
-              subscriptionService.subscribeAgencyToMtd(mapper(initialDetails.get, address)) flatMap { subscriptionResponse =>
-                sessionStoreService.remove() map { _ =>
-                  subscriptionResponse match {
-                    case Right(r) => Redirect(routes.SubscriptionController.showSubscriptionComplete())
-                      .flashing("arn" -> r.arn, "agencyName" -> mapper(initialDetails.get, address).name)
-                    case Left(CONFLICT) => Redirect(routes.CheckAgencyController.showAlreadySubscribed())
-                    case Left(FORBIDDEN) => Redirect(routes.SubscriptionController.showSubscriptionFailed())
-                    case Left(error) => InternalServerError(s"Unknown error code from agent-subscription $error")
-                  }
+              val subscriptionResponse = for {
+                detailsOpt <- sessionStoreService.fetchInitialDetails
+                subscriptionResponse <- detailsOpt match {
+                  case Some(details) => subscribe(details, address)
+                  case None => Future.successful(Left(MISSING_SESSION_DATA))
                 }
+                _ <- sessionStoreService.remove()
+              } yield subscriptionResponse
+
+              subscriptionResponse.map {
+                case Right((arn, agencyName)) => Redirect(routes.SubscriptionController.showSubscriptionComplete())
+                  .flashing("arn" -> arn.arn, "agencyName" -> agencyName)
+                case Left(CONFLICT) => Redirect(routes.CheckAgencyController.showAlreadySubscribed())
+                case Left(FORBIDDEN) => Redirect(routes.SubscriptionController.showSubscriptionFailed())
+                case Left(MISSING_SESSION_DATA) => Redirect(routes.SubscriptionController.showSubscriptionDetails())
```

Comment:
I'm not keen on the same val being able to contain two different things like this (either an HTTP status code returned by agent-subscription or a magic number to represent missing session data). It's unclear and error-prone.

How about using a sealed trait as the content of the Left. So `subscriptionResponse` would be a `Future[Either[SubscriptionFailed, (Arn, String)]]`.
```
  private sealed trait SubscriptionFailed
  private case class SubscriptionReturnedHttpError(httpStatusCode: Int) extends SubscriptionFailed
  private case object MissingSessionData extends SubscriptionFailed
```